### ISSUE TITLE
yubikey not working with KeePassXC 2.7.7 running on debian bookworm inside VirtualBox

### ISSUE THREAD

**Issue by the-wolfman** [keepassxreboot/keepassxc#10425]
## Overview
I am running KeePassXC on current debian bookworm. OS is installed inside a VirtualBox based virtual machine. My databases are secured with challenge-response to a yubikey 5 connected to the virtual machine via Virtualbox extension. This setup has been running for a number of years now up to and including KeePassXC 2.7.6. After the upgrade to 2.7.7, the yubikey gets detected, but the challenge response seems to hang infintely, the yubikey does not blink, the database stays locked. This happens with both the app image as well as the flatpak install. Everything is fine again after downgrading to 2.7.6.

## Steps to Reproduce
1. download and start KeePassXC inside a debian bookworm system running in Virtualbox
2. open existing yubikey secured database or create new database and add a yubikey
3. try to unlock
4. KeePassXC hangs until it receives a sigkill

## Expected Behavior
after step 3 above, the yubikey should start to blink. After touching the key, the database should be unlocked

## Actual Behavior
the yubikey is detected, correct serial number is shown, but the yubikey never starts to blink, KeePassXC hangs until sigkill

## Context
KeePassXC - Version 2.7.7
Revision: 68e2dd8
Distribution: AppImage

Qt 5.15.2
Diagnosemodus ist deaktiviert.

Betriebssystem: Debian GNU/Linux 12 (bookworm)
CPU-Architektur: x86_64
Kernel: linux 6.1.0-18-amd64

Aktivierte Erweiterungen:
- Auto-Type
- Browser-Integration
- Passkeys
- SSH-Agent
- KeeShare
- YubiKey
- Secret-Service-Integration

Kryptographische Bibliotheken:
- Botan 2.19.1

## Environment
Operating System: debian bookworm running in VirtualBox
Desktop Env: XFCE
Windowing System: X11

**Comment by droidmonkey** [comment]
@phoerious

**Comment by the-wolfman** [comment]
I wondered whether this may be related to an updated USB library packed into appimage and flatpak. So I tested the ppa version on an ubuntu 20.04 LTS virtual machine running on the same host. It shows exactly the same behavior. Can I help with providing additional info for debugging?

**Comment by droidmonkey** [comment]
Hmmmmm, I will test with my virtual box setup later today. I do my testing on Fedora

**Comment by the-wolfman** [comment]
Many thanks for looking into this!
I did some more tests and have the feeling it is somehow related to the UI flow, because this issue is only showing up at the unlock step. If I create a new database and then secure it with a yubikey challenge-resp, everything works as expected. The "blue ribbon" is showing up asking me to touch the yubikey and the database gets stored. When I then close KeePassXC, open it again and try to unlock, KeePassXC hangs before the blue ribbon shows or the yubikey blinks and most of he time does not react to sigterm but only sigkill. This is true not only for xfce, but also lxqt and gnome with wayland and x11, all on debian bookworm ... but not with Win10 as a guest system.
PLEASE NOTE: A few times I managed to overcome the issue. It's not strictly reproducable but when I already have another database openend (without chall-resp) and then try to open the database, the hang can sometimes be overcome by clicking onto the tab of the already opened database.
Not sure, frankly, how this might be related to Virtualbox, other than display drivers and the likes. I actually thought it should showup elsewhere as well. Thats where I came across [your comment here ](https://github.com/keepassxreboot/keepassxc/issues/10391#issuecomment-1995218093)

**Comment by droidmonkey** [comment]
Does the hang happen if you don't enable "touch required" for your yubikey challenge response slot?

**Comment by the-wolfman** [comment]
No difference, still hangs.
During the tests I repeatedly overcame the hang by clicking the tab of another database, openend or not, so this seems to be reproducable now.
Just a thought: While disabling "touch required" I recalled having issues years ago with access privileges/udev rules for yubikey OTP handled as HID device. Are you aware of changes in that area? I did some tests as root but did not notice any change.

**Comment by droidmonkey** [comment]
If you have udev problems you'll never see the key to begin with

**Comment by the-wolfman** [comment]
not sure, other features use pcsd for example, "seeing" the key and retrieving serial number while otp is not available. But I won't argue, because I agree in that case I would not be able to create a new database with yubikey support.

I also did test keepassxc-cli from the appimage on the same platform with no issues.

**Comment by the-wolfman** [comment]
I believe issues https://github.com/keepassxreboot/keepassxc/issues/10391 and https://github.com/keepassxreboot/keepassxc/issues/10425 are closely related or even duplicates.  @norbertj123 Very good catch.

Increasing the number of VCPUs from 2 to 3 or higher solved it for me. I thought at first it could be related to database encryption settings, specifically "threads", but I strongly believe that the issue is happening earlier. Reason is I encounter the issue specifically when securing the database with a yubikey and it is not even getting to the challenge-response.

Further testing shows
- I am having all kinds of issues when running with 1 VCPU
- I can open databases, but not one secured with a yubikey running with 2 VCPUs (worked earlier)
- Everything working fine - as far as I tested - running with 3 VCPUs or more

My take at the moment: This is a UI related race condition relying on threads to be available depending on features used. In my case, checking the yubikey and displaying the "please touch ribbon" starts a thread, or tries to, which won't come back as expected.

**Comment by droidmonkey** [comment]
Closing in favor of the other issue
